typescript/src/schema/proto.ts

```diff
@@ -17,10 +38,114 @@ interface MethodDescriptorProto
   inputInterface: string;
   outputInterface: string;
   comments: string;
+  methodConfig: plugin.grpc.service_config.MethodConfig;
+  retryableCodesName: string;
+  retryParamsName: string;
+  timeoutMillis?: number;
+}
+
+class RetryableCodeMap {
+  codeEnumMapping: { [index: string]: string };
+  uniqueCodesNamesMap: { [uniqueName: string]: string };
+  prettyCodesNamesMap: { [prettyName: string]: string[] };
+  uniqueParamsNamesMap: { [uniqueName: string]: string };
+  prettyParamNamesMap: { [prettyName: string]: {} };
+
+  constructor() {
+    this.uniqueCodesNamesMap = {};
+    this.prettyCodesNamesMap = {};
+    this.uniqueParamsNamesMap = {};
+    this.prettyParamNamesMap = {};
+
+    // build reverse mapping for enum: 0 => OK, 1 => CANCELLED, etc.
+    this.codeEnumMapping = {};
+    const allCodes = Object.keys(plugin.google.rpc.Code);
+    for (const code of allCodes) {
+      this.codeEnumMapping[
+        ((plugin.google.rpc.Code as unknown) as {
+          [key: string]: plugin.google.rpc.Code;
+        })[code].toString()
+      ] = code;
+    }
+
+    // generate some pre-defined code sets for compatibility with existing configs
+    this.getRetryableCodesName(
+      defaultNonIdempotentCodes,
+      defaultNonIdempotentRetryCodesName
+    );
+    this.getRetryableCodesName(
+      defaultIdempotentCodes,
+      defaultIdempotentRetryCodesName
+    );
+    this.getParamsName(defaultParameters, 'default');
+  }
+
+  private buildUniqueCodesName(
+    retryableStatusCodes: plugin.google.rpc.Code[]
+  ): string {
+    // generate an unique readable name for the given retryable set of codes
+    const sortedCodes = retryableStatusCodes.sort(
+      (a, b) => Number(a) - Number(b)
+    );
+    const uniqueName = sortedCodes
+      .map(code => this.codeEnumMapping[code])
+      .join('_')
+      .toSnakeCase();
```

Comment:
I might be misunderstanding this code, but it looks like it is concatenating the retry codes with a `_` to use as the name. I'm not sure that this name would actually be unique. `INTERNAL_UNAVAILABLE_UNKNOWN` as a code set name might not be unique.

Am I misunderstanding it?

Reply:
You are understanding it right. The idea is that I need some human readable name for an array of codes, e.g. `[UNAVAILABLE, INTERNAL, UNKNOWN]`, so I first sort them, then join with `_`.

It's indeed possible that the resulting code won't be unique if the `google.rpc.Code` enum has elements with `_`, e.g. (imaginary example) `UNKNOWN_ERROR`, `UNKNOWN`, and `ERROR`. So I looked at the actual contents of the enum and found that–if I'm not mistaken–the uniqueness of the code built by this method still holds:

```
{
  OK: 0,
  CANCELLED: 1,
  UNKNOWN: 2,
  INVALID_ARGUMENT: 3,
  DEADLINE_EXCEEDED: 4,
  NOT_FOUND: 5,
  ALREADY_EXISTS: 6,
  PERMISSION_DENIED: 7,
  UNAUTHENTICATED: 16,
  RESOURCE_EXHAUSTED: 8,
  FAILED_PRECONDITION: 9,
  ABORTED: 10,
  OUT_OF_RANGE: 11,
  UNIMPLEMENTED: 12,
  INTERNAL: 13,
  UNAVAILABLE: 14,
  DATA_LOSS: 15
}
```

It's technically possible that new codes will be added that will make this method generate non-unique codes but the possibility of it is very low in my opinion (and concatenating with `_` makes the keys readable).

Reply:
Ok so if two different MethodConfigs have the same set of retryable codes `[UNAVAILABLE, UNKNOWN]`, their RPCs would both use the same generated retry code set named `UNAVAILABLE_UNKNOWN` and that's OK because the set of the retryable codes between the two MethodConfigs are the same. Gotchya.